### PR TITLE
ap_pwrseq: Introduce Application Processor (AP) power sequencer driver

### PR DESCRIPTION
**Pull request by perezpri**
This driver will monitor well defined input signals to manage AP power
state transitions. AP power sequencer is designed to follow the zephyr
State Machine Framework (SMF). This driver is AP agnostic.

This initial commit shows the mechanism for its integration, it includes
example for architecture and chipset level power state transition handlers.

Signed-off-by: Bernardo Perez Priego <bernardo.perez.priego@intel.com>

### CODE REVIEW

File: dts/bindings/ap_pwrseq/ap-pwrseq-sub-states.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+ # Copyright 2022 The Chromium OS Authors. All rights reserved.
```
Comment (vphirema) [inline dts/bindings/ap_pwrseq/ap-pwrseq-sub-states.yaml:1]: intel copywrite?

File: include/zephyr/ap_pwrseq/ap_pwrseq.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		())
+
+#define AP_PWRSEQ_EACH_APP_SUB_STATE_ENUM_DEF(node_id)                        \
+	COND_CODE_1(DT_NODE_HAS_PROP(node_id, application),                   \
+		(DT_FOREACH_PROP_ELEM(node_id, application,                   \
+			AP_POWER_SUB_STATE_ENUM_DEF_WITH_COMMA)),             \
+		())
+
+/** @brief AP power sequence valid power states. */
+enum ap_pwrseq_state {
```
Comment (vphirema) [inline include/zephyr/ap_pwrseq/ap_pwrseq.h:34]: add a psudo G3 also,

File: include/zephyr/ap_pwrseq/ap_pwrseq.h [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	AP_POWER_STATE_G3,
```
Comment (vphirema) [inline include/zephyr/ap_pwrseq/ap_pwrseq.h:35]: add comments for each states regarding what is the intent

File: include/zephyr/ap_pwrseq/ap_pwrseq.h (hunk 2 above)
Comment (perezpri) [reply include/zephyr/ap_pwrseq/ap_pwrseq.h:34]: I am not aware of what this pseudo state is, is it UNINIT?

File: include/zephyr/ap_pwrseq/ap_pwrseq.h (hunk 3 above)
Comment (perezpri) [reply include/zephyr/ap_pwrseq/ap_pwrseq.h:35]: Ack

File: dts/bindings/ap_pwrseq/ap-pwrseq-sub-states.yaml (hunk 1 above)
Comment (perezpri) [reply dts/bindings/ap_pwrseq/ap-pwrseq-sub-states.yaml:1]: ACK

File: include/zephyr/ap_pwrseq/ap_pwrseq.h (hunk 2 above)
Comment (vphirema) [reply include/zephyr/ap_pwrseq/ap_pwrseq.h:35]: /* Both EC and AP are off */
AP_POWER_STATE_PG3

**Pull request closed without merging**